### PR TITLE
Onion client 14

### PR DESCRIPTION
**Pull request by CamiloGarciaLaRotta**
**Solving issue** #14

- Code in this branch was tested and meets the criteria of the open issue
- Minor changes in `onion_router.py` to avoid using low-level libs and adhere to the no JSON communication standard
- Minor changes in `dummy_dest.py` to reply to setup SYNs and adhere to the no JSON communication standard
- Folder structure was reformatted to homogenize naming conventions

**Logical Next Steps:** 
- Must actually implement directory's `get_path()` to return a valid JSON with path's hostnames and keys
- Must decide on whats the protocol of communication with the directory node. At this moment I send an empty `b'TODO'` when connecting to it, because the directory node doesn't care what the recieved message says.
- Must **DEFINETIVELY** handle teardown and error handling along the path: clossing one node crashes all other nodes in the VC

### CODE REVIEW

File: destination/dummy_dest.py [hunk 1]
```diff
@@ -41,9 +44,13 @@ def main():
 def reply(conn_socket):
     try:
         while True:
-            msg = conn_socket.recv(2048)
-            mod_msg = msg.upper()
-            conn_socket.send(mod_msg)
+            msg = conn_socket.recv(2048).decode('utf-8')
+            print('Recieved: {}'.format(msg))
+            if msg == 'SYN': 
+                conn_socket.send(b'ACK')
```
Comment (DestructiveReasoning) [inline destination/dummy_dest.py]: Why is this handshake necessary?

File: destination/dummy_dest.py [hunk 2]
```diff
@@ -41,9 +44,13 @@ def main():
 def reply(conn_socket):
     try:
         while True:
-            msg = conn_socket.recv(2048)
-            mod_msg = msg.upper()
-            conn_socket.send(mod_msg)
+            msg = conn_socket.recv(2048).decode('utf-8')
+            print('Recieved: {}'.format(msg))
+            if msg == 'SYN':
```
Comment (DestructiveReasoning) [inline destination/dummy_dest.py]: Why is this handshake necessary?

**Comment by DestructiveReasoning** [comment]
Looks nice. I just don't understand what that SYN/ACK thing is with the `dummy_dest.py`, and likewise the `received SYN` log in `onion_router.py`.

**Comment by CamiloGarciaLaRotta** [comment]
Not necessary, the same way the routers returning ACK is not necessary
right now.

I added it to keep consistency throughout the configuration of the VC.

You can see its use in the client's setup_vc()

File: client/onion_client.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+parser.add_argument("port", help="port to listen on", type=int)
+parser.add_argument("-v", "--verbose",
+                    help="level of logging verbose", action="store_true")
+
+args = parser.parse_args()
+
+# validate args against conditions
+if args.port < 5551 or args.port > 5557: # 7 group members, each get a port
+    parser.error("port must satisfy: 5551 <= port <= 5557")
+if args.verbose: print('Verbosity ON\n')
```
Comment (stumash) [inline client/onion_client.py]: I don't think it's conventional to tell the user via `print` statement that they have in fact passed the `-v` flag.

**Comment by CamiloGarciaLaRotta** [comment]
True, I can remove it

File: client/onion_client.py (hunk 3 above)
Comment (CamiloGarciaLaRotta) [reply client/onion_client.py]: True indeed, fixed in 267e7f2

File: destination/dummy_dest.py (hunk 2 above)
Comment (stumash) [reply destination/dummy_dest.py]: agree

File: directory/onion_directory.py [hunk 4]
```diff
@@ -110,14 +109,17 @@ def handle_path_request(conn):
...
     conn.close()
 
 
 def get_path():
-    return random.sample(NODES, 3)
+    #return random.sample(NODES, 3)
+    # TODO implement actual retrieval of data
+    return [{'addr':'lab2-1','key':'KEY_1'},
+            {'addr':'lab2-2','key':'KEY_2'},
+            {'addr':'lab2-3','key':'KEY_3'}]
```
Comment (stumash) [inline directory/onion_directory.py]: I don't see why this change was necessary

File: destination/dummy_dest.py (hunk 2 above)
Comment (CamiloGarciaLaRotta) [reply destination/dummy_dest.py]: @DestructiveReasoning @stumash 

To clarify on the SYN/ACK:

I had to adhere to the method of `onion_router.py` of establishing the VC.
Because each router returns ACK on the first message and then waits for the second one to establish the next hop connection, I had to choose a dummy string to send as a first message.
Choosing SYN was the natural answer

`onion_router.py` line 63: 
```python
#(TODO:decrypt data, initialize+respond with symkey. for now send 'ACK')
back_conn.sendall("ACK".encode('utf-8'))
```

As to why it was also added to the destination node, it was to keep consistency throughout the configuration of the VC as you can see here:

`onion_client.py` line 102
```python
 def setup_vc(path, conn):
 # get ACK from first node
 conn.sendall(b'SYN')
 response = conn.recv(1024).decode('utf-8').rstrip()
 if response != 'ACK': return false
 if args.verbose: print('ACK recieved')

 # get ACK for remaining nodes and destination
 for i in range(1,4):
     setup_onion = encapsulate(path[i]['addr'], path[i]['key'], 'SYN')
     if args.verbose: print('setup_onion: {}'.format(setup_onion))
     
     # TODO encrypt data
     
     conn.sendall(json.dumps(setup_onion).encode('utf-8'))
     response = conn.recv(1024).decode('utf-8').rstrip()
      if response != 'ACK': return false
     if args.verbose: print('ACK recieved')
```

I am aware it won't allow accessing websites because they won't return `ACK`s, I chose to add because currently, we are only talking to internal nodes. But if issue #15 is expected to be solved before the project submission and thus allowing us to communicate to other ports (e.g. accessing websites) I can change it right now.

I'll wait on your decision

Comment (DestructiveReasoning) [reply destination/dummy_dest.py]: I don't think this is necessary, because it's just not a realistic enough scenario in my opinion. The originator knows how many intermediate nodes there are, so can't he just wait for that many ACK's?

File: client/onion_client.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+def handle_response(conn):
+    while True:
+        msg = conn.recv(2048).decode('utf-8').rstrip()
+
+        # TODO decrypt
+
+        print('\nReply from {}: {}'.format('ADD SRC',msg))
+
+
+def shut_down_directory_node():
```
Comment (stumash) [inline client/onion_client.py]: *`shut_down_client_node` you mean?

File: client/onion_client.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
+import argparse             # for command-line argument parsing
+def shut_down_directory_node():
```
Comment (stumash) [inline client/onion_client.py]: this function is defined twice

File: client/onion_client.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+#################################################################################
+
+HOST = ''                            # empty string => all IPs of this machine
+PORT = args.port                     # port for server to listen on, cli arg
+
+DIRECTORY_NODE = 'cs-1'              # TODO hardcoded value or CL argument
+
+LISTEN_SOCKET = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+LISTEN_SOCKET.bind((HOST, PORT))
+LISTEN_SOCKET.listen(1)
```
Comment (stumash) [inline client/onion_client.py]: the `LISTEN_SOCKET` is not used, we don't need it

File: destination/dummy_dest.py (hunk 2 above)
Comment (stumash) [reply destination/dummy_dest.py]: I see. You put the handshake here so that you can do the loop in `setup_vc` in `onion_client`.  Also so that all nodes (whether it be directory, router, or destination) respond with ACK first before they then go on to do whatever it is that they do.

However, actual destinations (like, www.google.com, for example) don't follow this protocol - they just respond.  So, maybe it's not such a good idea for the destination to imitate the behaviour of the routers for 'consistency'.  Maybe, `setup_vc` should only initialise communication between the routers and that the first packet from the originator intended for processing by the destination will actually be the first packet that the destination receives from the originator -- as opposed to how in this PR, the destination receives a `SYN` as part of the `setup_vc`.

Do you agree?

File: directory/onion_directory.py [hunk 8]
```diff
@@ -5,11 +5,10 @@
 import threading            # for one thread per TCP connection
 
 
-
 #   TODO:
 #    - incorporate encryption key mgmt
-
-
+#    - directory ping of nodes causes routers to fail, must start directory
+#      before router
```
Comment (stumash) [inline directory/onion_directory.py:11]: this wasn't the case for me

File: directory/onion_directory.py (hunk 4 above)
Comment (CamiloGarciaLaRotta) [reply directory/onion_directory.py]: it is not a change, rather a stub based on the specifications for the directory node:
it should return 3 addresses and their keys.
Coding that functionality isn't part of the scope of this PR-issue

File: destination/dummy_dest.py (hunk 2 above)
Comment (CamiloGarciaLaRotta) [reply destination/dummy_dest.py]: that works, so to conclude:
- Recieve 3 `ACK`s because 3 intermediate nodes
- For the actual destination first package received is actual first data packet

When you confirm I'll apply the change

File: client/onion_client.py (hunk 6 above)
Comment (CamiloGarciaLaRotta) [reply client/onion_client.py]: Fixed in  2dadc05

File: client/onion_client.py (hunk 7 above)
Comment (CamiloGarciaLaRotta) [reply client/onion_client.py]: Fixed in  2dadc05

File: directory/onion_directory.py (hunk 8 above)
Comment (CamiloGarciaLaRotta) [reply directory/onion_directory.py:11]: it is, when the error occurred I rolled back to your latest change on `onion_router.py` 645c96cca878fa63bf5a07826560de6446f32856

It crashes there too, I assumed that's why you opened issue #19

File: client/onion_client.py (hunk 5 above)
Comment (CamiloGarciaLaRotta) [reply client/onion_client.py]: Fixed in 2dadc05

File: destination/dummy_dest.py (hunk 2 above)
Comment (stumash) [reply destination/dummy_dest.py]: yes cool

File: directory/onion_directory.py (hunk 4 above)
Comment (stumash) [reply directory/onion_directory.py]: it seems you changed the stub from doing `random.sample(NODES, 3)` to 
```python
return [{'addr':'lab2-1','key':'KEY_1'},
        {'addr':'lab2-2','key':'KEY_2'},
        {'addr':'lab2-3','key':'KEY_3'}]
```

but ya I guess it's just a formatting thing

File: directory/onion_directory.py (hunk 8 above)
Comment (stumash) [reply directory/onion_directory.py:11]: embarassing.  We'll have to handle that. nice find.

**Review by stumash — APPROVED** [review]
lookin good

File: destination/dummy_dest.py (hunk 2 above)
Comment (DestructiveReasoning) [reply destination/dummy_dest.py]: Yeah that's what I meant. Sounds good

**Review by DestructiveReasoning — APPROVED** [review]

File: destination/dummy_dest.py (hunk 2 above)
Comment (CamiloGarciaLaRotta) [reply destination/dummy_dest.py]: Fixed in 31e617e7b16ab2a1f0db2407a3bcd6f22e4cab33
